### PR TITLE
Functions for correlation matrix, copula indicator and plotting

### PR DESCRIPTION
**Pull request by SotirisTouliopoulos**
Aim of this PR is: 
  1. to provide functions for calculation of correlation matrix from steady states, filtering based on copula indicator & plotting
  2. to improve the efficiency of the `PreProcess` class.

The `correlated_reactions` function that calculates a pearson correlation matrix from reactions steady states is appended to the script `dingo/utils.py` Correlations that do not make the pearson cutoff are replaced with 0. For pairwise reactions with a greater pearson coefficient than the cutoff, a copula indicator is computed to filter false-positive correlations. Parameters of this function can adjust the width of the copula's diagonal for the calculation of the indicator and return only the lower triangle of the symmetric matrix. Cutoffs for pearson and indicator filtering can be adjusted from the user too.

The `plot_corr_matrix` function that creates a heatmap plot of a correlation matrix is appended to the script `dingo/illustrations.py` Parameters of this function can specify the format of the saved image plot and place as labels only the remained reactions for reduced models.

The `reduce` function of the `PreProcess` class for the `extend` parameter set to `True` has also changed to identify additional reactions for removal in bigger models. It uses the new `correlated_reactions` function to calculate a correlation matrix and order reactions based on their sum of absolute correlations with other reactions. Reactions with smaller values of overall correlation are removed first. The removal stops in the first reaction that alters the value of the objective function.

A `correlation.py` unittest is also created in the `tests` directory and the `preprocess.py` unittest is updated too.

### CODE REVIEW

**Review by TolisChal — APPROVED** [review]

File: dingo/utils.py [hunk 1]
```diff
@@ -201,3 +201,127 @@ def get_matrices_of_full_dim_polytope(A, b, Aeq, beq):
...
+            else:
+                filtered_corr_matrix[index1, index2] = 0
+                filtered_corr_matrix[index2, index1] = 0
+                
+            print("Completed process of",i+1,"from",corr_indices.shape[0],"copulas")
+
+                
+        print(positive, "out of", i+1, "copulas were positive correlated based on copula indicator")
+        print(negative, "out of", i+1, "copulas were negative correlated based on copula indicator")
+
```
Comment (TolisChal) [inline dingo/utils.py]: The prints should be requested by the user I think. These numbers could be part of the output as with a label for each copula: "positive" or "negative"

Comment (SotirisTouliopoulos) [reply dingo/utils.py]: I created a dictionary of this format that is returned

```
{'ACONTb_ACONTa': {'pearson': 0.9999999999999996, 'indicator': 700000000.9999999, 'classification': 'positive'},
 'ACt2r_ACKr': {'pearson': 1.0, 'indicator': 700000000.9999999, 'classification': 'positive'}}

File: dingo/utils.py [hunk 2]
```diff
@@ -201,3 +201,141 @@ def get_matrices_of_full_dim_polytope(A, b, Aeq, beq):
...
+            # value in the correlation matrix with 0
+            else:
+                filtered_corr_matrix[index1, index2] = 0
+                filtered_corr_matrix[index2, index1] = 0
+                pearson = filtered_corr_matrix[index1, index2]
+                indicator_dict[reaction1 + "~" + reaction2] = {'pearson': pearson,
+                                                               'indicator': indicator, 
+                                                               'classification': "no correlation"}
+                
+            print("Completed process of",i+1,"from",corr_indices.shape[0],"copulas")
```
Comment (TolisChal) [inline dingo/utils.py]: All the prints should ideally requested by the user and be off by default.
I'd use a flag `verbose: bool` as an input variable with `False` as default value. Then, print any message only if verbose is true.

File: dingo/preprocess.py [hunk 3]
```diff
@@ -209,50 +212,65 @@ def reduce(self, extend=False):
...
+                    finally:
+                        if (fba_solution_after != None) and (abs(fba_solution_after - fba_solution_before) < tol):
+                            self._removed_reactions.append(reaction)
+                            additional_removed_reactions_count += 1
+                        else:
+                            # restore bounds
+                            self._model.reactions.get_by_id(reaction).bounds = self._reaction_bounds_dict[reaction]
+
+              
+            print(len(self._removed_reactions), "of the", len(self._initial_reactions), \
```
Comment (TolisChal) [inline dingo/preprocess.py]: Plz, see my comment below for the printed messages.

**Pull request merged**